src/gsym/parser.rs

```
@@ -0,0 +1,330 @@
+//!
+//! See https://reviews.llvm.org/D53379
+use super::types::*;
+
+use std::io::{Error, ErrorKind};
+
+use crate::tools::{decode_udword, decode_uhalf, decode_uword};
+use std::ffi::CStr;
+
+/// Hold the major parts of a standalone GSYM file.
+///
+/// GsymContext provides functions to access major entities in GSYM.
+/// GsymContext can find respective AddressInfo for an address.  But,
+/// it doesn't parse AddressData to get line numbers.
+pub struct GsymContext<'a> {
+    header: Header,
+    addr_tab: &'a [u8],
+    addr_data_off_tab: &'a [u8],
+    str_tab: &'a [u8],
+    raw_data: &'a [u8],
+}
+
+impl<'a> GsymContext<'a> {
+    /// Parse the Header of a standalone GSYM file.
+    ///
+    /// # Arguments
+    ///
+    /// * `data` - is the content of a standalone GSYM.
+    ///
+    /// Returns a GsymContext, which includes the Header and other important tables.
+    pub fn parse_header<'d>(data: &'d [u8]) -> Result<GsymContext<'d>, Error> {
+        let mut off = 0;
+        // Parse Header
+        let magic = decode_uword(data);
+        if magic != GSYM_MAGIC {
+            return Err(Error::new(ErrorKind::InvalidData, "invalid magic number"));
+        }
+        off += 4;
+        let version = decode_uhalf(&data[off..]);
+        if version != GSYM_VERSION {
+            return Err(Error::new(ErrorKind::InvalidData, "unknown version number"));
+        }
+        off += 2;
+        let addr_off_size = data[off];
+        off += 1;
+        let uuid_size = data[off];
+        off += 1;
+        let base_address = decode_udword(&data[off..]);
+        off += 8;
+        let num_addrs = decode_uword(&data[off..]);
+        off += 4;
+        let strtab_offset = decode_uword(&data[off..]);
+        off += 4;
+        let strtab_size = decode_uword(&data[off..]);
+        off += 4;
+        let uuid: [u8; 20] = (&data[off..(off + 20)])
+            .try_into()
+            .expect("input data is too short");
```

It seems wrong to panic here. Why not return an error?